### DISCUSSION TITLE
SQL Injection Prevention | place holders | array

### DISCUSSION THREAD

**Discussion by mariobernheim** [ADOdb/ADOdb#1030] [Q&A] [answered]
Alright lads I hope i'm asking this in the right place, as its not an issue but a question :) the documentation regarding this sends you in different directions as there's quite a bit on the subject but examples all differ, maybe they're all different means to the same end but I thought I'd ask directly here in the community :)

```
$sql = "SELECT * from clients WHERE username = ? AND pwd = ? and status='Active' LIMIT 1";
$result = $db->execute($sql, array($username, md5($password+$salt)));
```

will this protect against sql injections rather than using the following from the documentation, also - using the above, would it make it unnecessary to use qstr(); or addq();

```
$p1 = $db->param('p1');
$p2 = $db->param('p2');
```
the database is MySQL

**Comment by dregad** [comment]
Using parameterized queries is the correct way to prevent SQL injection.

Inserting hardcoded `?` placeholders in your SQL statement as per your example is fine if you do not care about portability. If you do then you should use *[param()](https://adodb.org/dokuwiki/doku.php?id=v5:reference:connection:param)* instead, like so:

```php
$sql = "SELECT * from clients WHERE username = " . $db->param('username') 
    . " AND pwd =  " . $db->param('hash') . " and status='Active' LIMIT 1";
```

You always need to provide the placeholders' values as an array to *execute()*'s 2nd argument.

I would not recommend to use _addQ()_ or _qStr()_; while they should effectively protect you from SQL injection, they also prevent you from using prepared statements, which could negatively impact performance if the same query has to be executed multiple times.

As a side note, MD5 to store a password, really !? Consider using [password_hash()](https://www.php.net/manual/en/function.password-hash.php) instead.

**Reply by mariobernheim** [reply]
Thanks so much , this is an inherited code we're hardening, the md5 is actually tomorrow's bridge we'll need to find a way to migrate existing clients.

Are the hardcoded ` ?` binding placeholders in the queries available in v5.20.17? We'll be migrating to the newest release but want make sure we at least have the most forward facing services hardened against SQL Injection.

**Reply by dregad** [reply, marked as answer]
Yes, the `?` placeholders have been there for as long as I have been using ADOdb.

**Reply by dregad** [reply]
For the record, please note that the mysql driver < 5.22.0 is not using "true" prepared statements with bind parameters (see #655), only emulated ones using qstr() but still you should be safe. I recommend to upgrade as soon as possible.
